lisa/analysis/cpus.py

```diff
@@ -66,6 +66,96 @@ def df_context_switches(self):
 
         return ctx_sw_df
 
+    @TraceAnalysisBase.cache
+    @requires_events('sched_switch')
+    def df_states(self):
+        """
+        Compute the state intervals on each CPU.
+
+        :returns: A :class:`pandas.DataFrame` with:
+
+          * A ``cpu`` column (the CPU the state refers to)
+          * A ``state`` column (the BUSY/IDLE state the CPU is into)
+          * A ``duration`` column (the time the CPU is in this state)
+          * A ``end_time`` column (the time the CPU will exit this state)
+        """
+        # Start from sched_switch events
+        sched_df = self.trace.df_event('sched_switch')
+        
+        # Keep only CPU transition events (IDLE to/from BUSY)
+        def switch_cpu_state(row):
+            if row.prev_pid != 0 and row.next_pid != 0:
+                return False
+            return True
+        states_df = sched_df[sched_df.apply(switch_cpu_state, axis=1)]
+        
+        # Reset index and use the event timestamp to compute deltas
+        states_df.reset_index(inplace=True)
+        
+        # Compute next transition time and deltas (by grouping events by CPU)
+        grouped = states_df.groupby('__cpu', observed=True, sort=False)
+        new_columns = dict(
+            end_time=grouped['Time'].shift(-1),
+            # GroupBy.transform() will run the function on each group, and
+            # concatenate the resulting series to create a new column.
+            # Note: We actually need transform() to chain 2 operations on
+            # the group, otherwise the first operation returns a final
+            # Series, and the 2nd is not applied on groups
+            duration=grouped['Time'].transform(lambda time: time.diff().shift(-1)),
```

Comment:
Use `lisa.datautils.df_add_delta(window=self.trace.window)`. 
`time.diff().shift(-1)` will create a `NaN` duration on the last row which is incorrect and can become a real issue in some circumstances (e.g. a when slicing on the trace to get a very small window)